src/Ormolu/Printer/Meat/Declaration/Value.hs

```diff
@@ -971,28 +971,56 @@ p_if ::
   (body -> Placement) ->
   -- | Render
   (body -> R ()) ->
+  -- | Annotations
+  EpAnn AnnsIf ->
   -- | If
   LHsExpr GhcPs ->
   -- | Then
   LocatedA body ->
   -- | Else
   LocatedA body ->
   R ()
-p_if placer render if' then' else' = do
+p_if placer render epAnn if' then' else' = do
   txt "if"
   space
   located if' p_hsExpr
   breakpoint
   inci $ do
-    txt "then"
+    locatedToken mThenSpan "then"
     space
-    located then' $ \x ->
-      placeHanging (placer x) (render x)
+    placeHangingLocated mThenSpan then'
     breakpoint
-    txt "else"
+    locatedToken mElseSpan "else"
     space
-    located else' $ \x ->
-      placeHanging (placer x) (render x)
+    placeHangingLocated mElseSpan else'
+  where
+    (mThenSpan, mElseSpan, mCommentSpans) =
```

Comment:
Currently, these have types
```haskell
mThenSpan, mElseSpan :: Maybe RealSrcSpan
mCommentSpans :: Maybe [RealSrcSpan]
```
What about changing this to
```haskell
thenSpan, elseSpan :: SrcSpan
commentSpans :: [RealSrcSpan]
```
This allows some minor simplifications below. As a replacement for `Nothing :: Maybe RealSrcSpan`, you can use eg [`noSrcSpan`](https://hackage.haskell.org/package/ghc-lib-parser-9.6.4.20240109/docs/GHC-Types-SrcLoc.html#v:noSrcSpan).

Reply:
hm I'm open, but I'm not sure how `mThenSpan :: SrcSpan` would simplify it. Can you elaborate?

e.g. the primary operation I need is `thenSpan < commentSpan`. But `commentSpan` is already a `RealSrcSpan`, so I might as well keep `thenSpan` a `RealSrcSpan`. Plus, `SrcSpan` doesn't have an `Ord` instance

Reply:
Eg the diff would look like this (should also remove the `m` prefix of `mThenSpan` and `mElseSpan`):
```diff
diff --git a/src/Ormolu/Printer/Meat/Declaration/Value.hs b/src/Ormolu/Printer/Meat/Declaration/Value.hs
index 1c65bba..5cc1ffb 100644
--- a/src/Ormolu/Printer/Meat/Declaration/Value.hs
+++ b/src/Ormolu/Printer/Meat/Declaration/Value.hs
@@ -997,28 +997,26 @@ p_if placer render epAnn if' then' else' = do
     (mThenSpan, mElseSpan, commentSpans) =
       case epAnn of
         EpAnn {anns = AnnsIf {aiThen, aiElse}, comments} ->
-          ( Just $ epaLocationRealSrcSpan aiThen,
-            Just $ epaLocationRealSrcSpan aiElse,
+          ( loc' $ epaLocationRealSrcSpan aiThen,
+            loc' $ epaLocationRealSrcSpan aiElse,
             map (anchor . getLoc) $
               case comments of
                 EpaComments cs -> cs
                 EpaCommentsBalanced pre post -> pre <> post
           )
         EpAnnNotUsed ->
-          (Nothing, Nothing, [])
+          (noSrcSpan, noSrcSpan, [])
 
-    locatedToken mTokenSpan token =
-      case mTokenSpan of
-        Just tokenSpan -> located (L tokenSpan ()) $ \_ -> txt token
-        Nothing -> txt token
+    locatedToken tokenSpan token =
+      located (L tokenSpan ()) $ \_ -> txt token
 
-    placeHangingLocated mTokenSpan bodyLoc@(L _ body) = do
+    placeHangingLocated tokenSpan bodyLoc@(L _ body) = do
       let hasComments = fromMaybe False $ do
-            tokenSpan <- mTokenSpan
+            realTokenSpan <- srcSpanToRealSrcSpan $ tokenSpan
             bodySpan <- srcSpanToRealSrcSpan $ getLoc' bodyLoc
-            pure $ any (\commentSpan -> tokenSpan < commentSpan && commentSpan < bodySpan) commentSpans
+            pure $ any (\commentSpan -> realTokenSpan < commentSpan && commentSpan < bodySpan) commentSpans
           placement = if hasComments then Normal else placer body
-      switchLayout ([loc' tokenSpan | Just tokenSpan <- pure mTokenSpan] <> [getLoc' bodyLoc]) $
+      switchLayout [tokenSpan, getLoc' bodyLoc] $
         placeHanging placement (located bodyLoc render)
 
 p_let ::
```
But this is a minor thing, feel free to also leave it as is :+1:

Reply:
ah perfect, thanks!